mbs_results/pivot_imputation_value.py

```diff
@@ -0,0 +1,169 @@
+    forward: str,
+    backward: str,
+    construction: str,
+    question: str,
+    imputed_value: str,
+    f_count: str,
+    b_count: str,
+    c_count: str,
+    selected_periods: list = None,
+) -> pd.DataFrame:
+
+    """
+    Returning dataframe containing imputation_value, filtered by date, pivoted by
+    imputation type and grouped by sic, cell, question and imputation type.
+
+    Parameters
+    ----------
+    dataframe : pd.DataFrame
+        Reference dataframe containing links, count values, and imputed values
+        by identifier, cell, date, and question
+    identifier : str
+        name of column in dataframe containing identifier variable
+    date : str
+        name of column in dataframe containing date variable
+    sic : str
+        name of column in dataframe containing sic variable
+    cell : str
+        name of column in dataframe containing cell variable
+    forward : str
+        name of column in dataframe containing forward link variable
+    backward : str
+        name of column in dataframe containing backward link variable
+    construction : str
+        name of column in dataframe containing construction link variable
+    question : str
+        name of column in dataframe containing question code variable
+    imputed_value: str
+        name of column in dataframe containing imputed_value variable
+    f_count: str,
+        name of column in dataframe containing f_count variable
+    b_count: str,
+        name of column in dataframe containing b_count variable
+    c_count: str,
+        name of column in dataframe containing c_count variable
+    selected_periods: list,
+        list containing periods to include in output
+
+    Returns
+    -------
+    dataframe filtered by date, containing imputation_value, pivoted by imputation type
+    and grouped by sic, cell, question and imputation type.
+
+    """
+    if selected_periods is not None:
+        df = df.query("{} in {}".format(date, selected_periods))
+
+    links_df = df.melt(
+        id_vars=[date, sic, cell, question, imputed_value],
+        value_vars=[forward, backward, construction],
+        var_name="link_type",
+        value_name="imputation_link",
+    )
+
+    link_type_map = {forward: "F", backward: "B", construction: "C"}
+    links_df["link_type"] = links_df["link_type"].map(link_type_map)
+
+    counts_df = df.melt(
+        id_vars=[date, sic, cell, question],
+        value_vars=[f_count, b_count, c_count],
+        var_name="link_type_count",
+        value_name="count",
+    )
+
+    link_type_map_count = {f_count: "F", b_count: "B", c_count: "C"}
+    counts_df["link_type_count"] = counts_df["link_type_count"].map(link_type_map_count)
+
+    merged_df = pd.merge(
+        links_df,
+        counts_df,
+        how="outer",
+        left_on=[date, sic, cell, question, "link_type"],
+        right_on=[date, sic, cell, question, "link_type_count"],
+    )
+
+    merged_df.drop_duplicates(inplace=True)
```

Comment:
Can we please add a comment why this is needed ?

Reply:
We've included the drop_duplicates command to prevent incorrect summed values - please let us know if you think another approach would be better.